scripts/generate_refcats_gen3.py

```diff
@@ -0,0 +1,248 @@
+import sys
+
+from esutil import htm
+from astropy.coordinates import SkyCoord
+
+import lsst.log
+from lsst.daf.butler import Butler, CollectionType, DatasetRef, DatasetType, FileDataset
+
+
+logging.basicConfig(level=logging.INFO, stream=sys.stdout)
+lsst.log.configure_pylog_MDC("DEBUG", MDC_class=None)
+
+
+########################################
+# Fields and catalogs to process
+
+FIELDS = [SkyCoord("10h00m28.800s", "+02d12m36.00s"),  # field 26 (2014:04)
+          SkyCoord("10h20m28.800s", "-06d31m12.00s"),  # field 40 (2014:10)
+          SkyCoord("10h21m52.800s", "-04d57m00.00s"),  # field 42 (2014:09)
+          ]
+FIELD_RADIUS = 2.0  # degrees
+
+# Key is catalog name in source repo, value is name in ap_verify_hits2015.
+REFCATS = {"gaia_dr2_20200414": "gaia",
+           "ps1_pv3_3pi_20170110": "panstarrs",
+           }
+
+
+########################################
+# Command-line options
+
+def _make_parser():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("-b", dest="src_dir", default="/repo/main",
+                        help="Refcat source Butler repo, defaults to '/repo/main'.")
+    parser.add_argument("-i", dest="src_collection", default="refcats",
+                        help="Refcat source collection, defaults to 'refcats'.")
+    return parser
+
+
+args = _make_parser().parse_args()
+
+
+########################################
+# Identify all required shards
+
+HTM_LEVEL = 7
+
+
+def _get_shards(centers, radius):
+    """Return all shards overlapping a set of fields.
+
+    Parameters
+    ----------
+    centers : iterable [`astropy.coordinates.SkyCoord`]
+        The right ascension and declination of the field centers.
+    radius : `float`
+        The radius of each field, in degrees.
+
+    Returns
+    -------
+    shards : `set` [`int`]
+        The set of HTM indices that overlap any of the fields.
+    """
+    indexer = htm.HTM(depth=HTM_LEVEL)
+    shards = set()
+    for center in centers:
+        shards.update(indexer.intersect(center.ra.degree, center.dec.degree, radius, inclusive=True))
+    return shards
+
+
+logging.info("Identifying refcat shards...")
+shards = _get_shards(FIELDS, FIELD_RADIUS)
+if not shards:
+    raise RuntimeError("No HTM shards found; coordinates are likely corrupted.")
+logging.debug("%d shards found", len(shards))
+
+
+# Simplify shard list so that Butler doesn't choke on a long query string.
+
+def _make_range(start, end):
+    """Represent a range of contiguous integers in Butler dimension
+    expression syntax.
+
+    Parameters
+    ----------
+    start, end : `int`
+        The first and last elements of the range, *inclusive*.
+        Assumes ``start <= end``.
+
+    Returns
+    -------
+    range : `str`
+        The Butler syntax for the range.
+    """
+    if start == end:
+        return str(start)
+    else:
+        return f"{start}..{end}"
+
+
+def _to_ranges(ids):
+    """Reexpress a set of integer IDs in terms of Butler dimension
```

Comment:
These two are useful methods, and I think could be useful in daf_butler, maybe, since it's useful for packing a large query. That would let us have some tests for it too: I can't quite convince myself on the loop logic.

Reply:
Can you explain what you see this being used for? I see it as an ultra-specific hack that I'm using because I couldn't wait for built-in spatial query support, and absolutely not something that should be emulated when you have time to get something cleaner.

As for correctness, I've run tests, and I also have a proof. I decided to spare readers the burden of having to scroll past the latter...

Reply:
Ah, I wasn't aware about what this was hacking around. You're good.